src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -0,0 +1,282 @@
+   * If the size is == 0, indicates that enqueued elements was consumed right away, otherwise it indicates
+   * size of the queue.
+   * Please note this will get completed _after_ `a` has been successfully enqueued.
+   * In case queue is `bounded` this waits to complete until queue is drained so there is space to enqueue `a`
+   * @param a `A` to enqueue
+   */
+  def enqueueOne(a: A): F[Int]
+
+  /**
+   * Enqueue multiple `A` values in this queue. This has same semantics as sequencing
+   * repeated calls to `enqueueOne`.
+   */
+  def enqueueAll(xa: Seq[A]): F[Int]
+
+  /**
+   * Provides a process that dequeue from this queue.
+   * When multiple consumers dequeue from this queue,
+   * they dequeue in first-come, first-serve order.
+   *
+   * Please use `Topic` instead of `Queue` when all subscribers
+   * need to see each value enqueued.
+   *
+   * This process is equivalent to `dequeueBatch(1)`.
+   */
+  def dequeue: Stream[F, A]
+
+  /**
+   * Provides a process that dequeues in chunks.  Whenever *n* elements
+   * are available in the queue, `min(n, limit)` elements will be dequeud
+   * and produced as a single `Chunk`.  Note that this naturally does not
+   * *guarantee* that `limit` items are returned in every chunk.  If only
+   * one element is available, that one element will be returned in its own
+   * Chunk.  This method basically just allows a consumer to "catch up"
+   * to a rapidly filling queue in the case where some sort of batching logic
+   * is applicable.
+   */
+  def dequeueBatch(limit: Int): Stream[F, A]
+
+  /**
+   * Equivalent to dequeueBatch with an infinite limit.  Only use this
+   * method if your underlying algebra (`A`) has some sort of constant
+   * time "natural batching"!  If processing a chunk of size n is linearly
+   * more expensive than processing a chunk of size 1, you should always
+   * use dequeueBatch with some small limit, otherwise you will disrupt
+   * fairness in the nondeterministic merge combinators.
+   */
+  def dequeueAvailable: Stream[F, A]
+
+  /**
+   * The time-varying size of this `Queue`. This signal refreshes
+   * only when size changes. Offsetting enqueues and de-queues may
+   * not result in refreshes.
+   */
+  def size: Stream[F,immutable.Signal[F,Int]]
+
+  /**
+   * The size bound on the queue.
+   * Returns None if the queue is unbounded.
+   */
+  def upperBound: Option[Int]
+
+  /**
+   * Returns the available number of entries in the queue.
+   * Always returns `Int.MaxValue` when the queue is unbounded.
+   */
+  def available: Stream[F,immutable.Signal[F,Int]]
+
+  /**
+   * Returns `true` when the queue has reached its upper size bound.
+   * Always returns `false` when the queue is unbounded.
+   */
+  def full: Stream[F,immutable.Signal[F,Boolean]]
+
+
+
+
+}
+
+object Queue {
+
+    private type DeQueue[F[_],A] = (Int,Seq[A] => F[Unit]) // Int represent preferred chunk size
```

Comment:
If you just make this A => F[Unit] and likewise for enqueue, it seems like it could simplify your implementation a lot.